### PR TITLE
Feature/405 extend stan variables

### PR DESCRIPTION
**Pull request by mitzimorris**
#### Submission Checklist

- [x] Run unit tests
- [x] Declare copyright holder and open-source license: see below

#### Summary

This PR generalizes the `stan_variables` function to optimize and variational methods, which requires the following:

- expose the `InferenceMedatadata` class
- populate CmdStanMCMC, CmdStanMLE, and CmdStanVB objects attribute `_metadata` during object initialization
- expose attr `_metadata` as property `metadata`
- generalize the term `sampler` to `method` when referring to output columns which end in `__`

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):  Columbia University

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)

### CODE REVIEW

**Comment by mitzimorris** [comment]
@ahartikainen and @WardBrian - this PR still needs to be extended to the output of generated quantities.

the question is how much refactoring to do in one PR.  right now, this adds a lot of pretty similar code to the classes that wrap a runset, i.e. CmdStanMCMC et al.

things could be simplified by simply exposing the "InferenceMetadata" object via method `metadata` on all the objects.

we might also be able to take things one step further by generalizing `stan_variable` and `stan_variables` to work with either a draws array or a vector of per-parameter estimates.

also there's this issue:  https://github.com/stan-dev/cmdstanpy/issues/407

what do y'all think?

File: cmdstanpy/stanfit.py [hunk 1]
```diff
@@ -1116,6 +1115,28 @@ def optimized_params_dict(self) -> OrderedDict:
         """Returns optimized params as Dict."""
         return OrderedDict(zip(self.column_names, self._mle))
 
+    def stan_variable(self, name: str) -> np.ndarray:
```
Comment (WardBrian) [inline cmdstanpy/stanfit.py:1118]: Similar to the existing `stan_variable` in the MCMC object, these new methods will want fairly detailed docstrings, describing at least the dimensionality of the return type.

File: cmdstanpy/stanfit.py [hunk 2]
```diff
@@ -485,12 +488,8 @@ def column_names(self) -> Tuple[str, ...]:
         and quantities of interest. Corresponds to Stan CSV file header row,
         with names munged to array notation, e.g. `beta[1]` not `beta.1`.
         """
-        if not self._validate_csv and self._metadata is None:
-            self._logger.warning(
-                'csv files not yet validated, run method validate_csv_files()'
-                ' in order to retrieve sample metadata.'
-            )
-            return None
+        if self._metadata is None:
```
Comment (WardBrian) [inline cmdstanpy/stanfit.py:491]: Even if we don't go the extra step of calling validate_csv_files in the constructors, I think these changes are great -- unexpectedly returning None can be a real headache

File: cmdstanpy/stanfit.py [hunk 3]
```diff
@@ -1323,6 +1371,28 @@ def variational_params_dict(self) -> OrderedDict:
...
+            raise ValueError('unknown name: {}'.format(name))
+        col_idxs = list(self._metadata.stan_vars_cols[name])
+        vals = list(self._variational_mean)
+        xs = [vals[x] for x in col_idxs]
+        shape = ()
+        if len(col_idxs) > 0:
+            shape = self._metadata.stan_vars_dims[name]
+        return np.ndarray(
+            shape=shape, buffer=np.array(xs)
+        )
```
Comment (WardBrian) [inline cmdstanpy/stanfit.py:1385]: I've never seen this form of creating an ndarray before - does it have an advantage over `np.array(xs).reshape(shape)`? Merely curious - the docs seem to suggest `np.ndarray` is primarily an internal constructor, and my assumption is it does two allocations?

File: cmdstanpy/stanfit.py [hunk 4]
```diff
@@ -509,61 +508,47 @@ def num_unconstrained_params(self) -> int:
...
-                'csv files not yet validated, run method validate_csv_files()'
-                ' in order to retrieve sample metadata.'
-            )
-            return None
+        if self._metadata is None:
+            self.validate_csv_files()
         return self._metadata.sampler_vars_cols
 
+    @property
+    def stan_vars_cols(self) -> Dict:
```
Comment (WardBrian) [inline cmdstanpy/stanfit.py:532]: I realize that there is a large body of code in the package which _doesn't_ do this, but the type hint convention is to specify the types in the dictionary (or other collection). For example, I believe the full return type of this function would be `Dict[str, Tuple[int]]`. This may be worth a separate PR to fix up all the existing types, which I'd be happy to do if we wanted

**Review by WardBrian — COMMENTED** [review]
This looks very useful to me! I made a few remarks:

File: cmdstanpy/stanfit.py (hunk 3 above)
Comment (mitzimorris) [reply cmdstanpy/stanfit.py:1385]: this is solely due to my lack of fluency in Python.  thanks for catching this.

File: cmdstanpy/utils.py [hunk 5]
```diff
@@ -672,7 +672,8 @@ def parse_method_vars(names: Tuple[str]) -> Dict[str, Tuple[int]]:
     return {v: tuple([k]) for (k, v) in enumerate(names) if v.endswith('__')}
 
 
-def parse_stan_vars(names: Tuple[str]) -> (Dict[str, Tuple[int]], Dict[str, Tuple[int]]):
+def parse_stan_vars(names: Tuple[str]) ->
+    (Dict[str, Tuple[int]], Dict[str, Tuple[int]]):
```
Comment (WardBrian) [inline cmdstanpy/utils.py]: I believe this should be 

`Tuple[Dict[str, Tuple[int,...], Dict[str, Tuple[int,...]` 

`Tuple[int]` is the type of a tuple with exactly one integer element, the ... makes it variable length of all ints. Ugly, I know. 

It may be worth adding something like mypy to the CI just to catch our own declarations

Comment (mitzimorris) [reply cmdstanpy/utils.py]: mypy found 139 errors.  will try not to introduce new ones in this PR.

Comment (WardBrian) [reply cmdstanpy/utils.py]: I can go through and address existing ones after this PR

**Comment by mitzimorris** [comment]
OK, this is ready for code review.  are the existing unit tests sufficient?
doing CmdStanGQ next.

File: cmdstanpy/stanfit.py (hunk 3 above)
Comment (WardBrian) [reply cmdstanpy/stanfit.py:1388]: This still needs to be resolved here and in the other new functions

File: cmdstanpy/stanfit.py [hunk 6]
```diff
@@ -1116,6 +1118,36 @@ def optimized_params_dict(self) -> OrderedDict:
...
+            raise ValueError('unknown name: {}'.format(name))
+        col_idxs = list(self._metadata.stan_vars_cols[name])
+        vals = list(self._mle)
+        xs = [vals[x] for x in col_idxs]
+        shape = ()
+        if len(col_idxs) > 0:
+            shape = self._metadata.stan_vars_dims[name]
+        return np.ndarray(
+            shape=shape, buffer=np.array(xs)
+        )
```
Comment (WardBrian) [inline cmdstanpy/stanfit.py:1139]: This should also change to `np.array(xs).reshape(shape)`

Comment (mitzimorris) [reply cmdstanpy/stanfit.py:1139]: done!

File: cmdstanpy/stanfit.py [hunk 7]
```diff
@@ -1135,7 +1135,7 @@ def stan_variable(self, name: str) -> np.ndarray:
         if len(col_idxs) > 0:
             shape = self._metadata.stan_vars_dims[name]
         return np.ndarray(
```
Comment (WardBrian) [inline cmdstanpy/stanfit.py:1137]: There's no need to wrap a call to np.array in np.ndarray, so the full return statement for both of these should just be `return np.array(xs).reshape(shape)`

**Review by WardBrian — CHANGES_REQUESTED** [review]
The tests any everything look good to me, just another minor nitpick

File: cmdstanpy/stanfit.py [hunk 8]
```diff
@@ -1134,9 +1132,10 @@ def stan_variable(self, name: str) -> np.ndarray:
         shape = ()
         if len(col_idxs) > 0:
             shape = self._metadata.stan_vars_dims[name]
-        return np.ndarray(
-            np.array(xs).reshape(shape)
-        )
+        print('xs: {}'.format(xs))
+        print('len xs: {}'.format(len(xs)))
+        print('shape: {}'.format(shape))
```
Comment (WardBrian) [inline cmdstanpy/stanfit.py:1137]: This seems like debugging code that was left in

Comment (mitzimorris) [reply cmdstanpy/stanfit.py]: fixed

**Review by WardBrian — APPROVED** [review]
This looks good to me now
